gnpy/core/elements.py

```diff
@@ -417,6 +417,201 @@ def __call__(self, spectral_info):
+        """
+        # TODO|jla: discuss beta2 as method or attribute
+        wl = 1550e-9 if ref_wavelength is None else ref_wavelength
+        D = abs(self.dispersion)
+        b2 = (wl ** 2) * D / (2 * pi * c)  # 10^21 scales [ps^2/km]
+        return b2 # s/Hz/m
+
+    def dbkm_2_lin(self):
+        """ calculates the linear loss coefficient
+        """
+        # alpha_pcoef is linear loss coefficient in dB/km^-1
+        # alpha_acoef is linear loss field amplitude coefficient in m^-1
+        alpha_pcoef = self.loss_coef
+        alpha_acoef = alpha_pcoef / (2 * 10 * log10(exp(1)))
+        return alpha_pcoef, alpha_acoef
+
+    def _psi(self, carrier, interfering_carrier):
+        """ Calculates eq. 123 from arXiv:1209.0394.
+        """
+        if carrier.num_chan == interfering_carrier.num_chan: # SCI
+            psi = arcsinh(0.5 * pi**2 * self.asymptotic_length
+                              * abs(self.beta2()) * carrier.baud_rate**2)
+        else: # XCI
+            delta_f = carrier.freq - interfering_carrier.freq
+            psi = arcsinh(pi**2 * self.asymptotic_length * abs(self.beta2())
+                                * carrier.baud_rate * (delta_f + 0.5 * interfering_carrier.baud_rate))
+            psi -= arcsinh(pi**2 * self.asymptotic_length * abs(self.beta2())
+                                 * carrier.baud_rate * (delta_f - 0.5 * interfering_carrier.baud_rate))
+
+        return psi
+
+    def _gn_analytic(self, carrier, *carriers):
+        """ Computes the nonlinear interference power on a single carrier.
+        The method uses eq. 120 from arXiv:1209.0394.
+        :param carrier: the signal under analysis
+        :param carriers: the full WDM comb
+        :return: carrier_nli: the amount of nonlinear interference in W on the under analysis
+        """
+
+        g_nli = 0
+        for interfering_carrier in carriers:
+            psi = self._psi(carrier, interfering_carrier)
+            g_nli += (interfering_carrier.power.signal/interfering_carrier.baud_rate)**2 \
+                     * (carrier.power.signal/carrier.baud_rate) * psi
+
+        g_nli *= (16 / 27) * (self.gamma * self.effective_length)**2 \
+                 / (2 * pi * abs(self.beta2()) * self.asymptotic_length)
+
+        carrier_nli = carrier.baud_rate * g_nli
+        return carrier_nli
+
+    def update_pref(self, pref, *carriers):
+        pch_out_db = 10*log10(mean([carrier.power.signal for carrier in carriers])) + 30
+        self.pch_out_db = round(pch_out_db, 2)
+        return pref._replace(p_span0=pref.p0, p_spani=self.pch_out_db)
+
+    def __call__(self, spectral_info):
+        self.carriers_in = spectral_info.carriers
+        carriers = tuple(self.propagate(*spectral_info.carriers))
+        pref = self.update_pref(spectral_info.pref, *carriers)
+        self.carriers_out = carriers
+        return spectral_info.update(carriers=carriers, pref=pref)
+
+    def propagate(self, *carriers):
```

Comment:
Apart from a small difference in the constructor, this `propagate` method (and one difference above) is the only difference between classes `Fiber` and `RamanFiber` (as long as I'm reading the diff correctly, accounting for some cleanup in the develop branch). I would prefer to have `RamanFiber` derived from `Fiber`, overriding the `propagate` method and the constructor. This will reduce code duplication; in general, copying code around is something that should be avoided.